File: Soutenance 2/Miny-RPG Editor/Mainfrm.vb
Public Class Mainfrm

    Public Map As New Map
    Private Painting As Boolean = False
    Private FileName As String = ""
    Public EditionType As Integer = 0

    Private LastCell As Integer = -1

    Private Sub Mainfrm_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        InitMap()
    End Sub

    Private Sub NouveauToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles NouveauToolStripMenuItem.Click
        InitMap()
    End Sub

    Private Sub QuitterToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles QuitterToolStripMenuItem.Click
        ExitProgram()
    End Sub

    Private Sub GrilleToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles GrilleToolStripMenuItem.Click
        DrawAll()
    End Sub

    Private Sub Couche1ToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Couche1ToolStripMenuItem.Click
        DrawAll()
    End Sub

    Private Sub Couche2ToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Couche2ToolStripMenuItem.Click
        DrawAll()
    End Sub

    Private Sub Couche3ToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Couche3ToolStripMenuItem.Click
        DrawAll()
    End Sub

    Private Sub MainPicture_MouseDown(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs) Handles MainPicture.MouseDown
        Painting = True
    End Sub

    Private Sub MainPicture_MouseUp(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs) Handles MainPicture.MouseUp
        Painting = False
    End Sub

    Private Sub MainPicture_MouseMove(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs) Handles MainPicture.MouseMove
        Dim X As Integer = Math.Floor(e.X / TILE_WIDTH)
        Dim Y As Integer = Math.Floor(e.Y / TILE_WIDTH)

        If Not Map.IsOnMap(X, Y) Then Exit Sub

        Dim Cell As Integer = Cells.GetId(X, Y)

        If Cell <> LastCell Then
            LastCell = Cell

            LabelStatus.Text = "X: " & X & "  Y: " & Y & "  Id: " & Cell

            If EditionType = 0 Then

                If Painting AndAlso Modfrm.BrushContinue.Checked Then

                    If e.Button = Windows.Forms.MouseButtons.Left Then
                        PaintOnCell(Cell, False)
                    ElseIf e.Button = Windows.Forms.MouseButtons.Right Then
                        PaintOnCell(Cell, True)
                    End If

                End If

            End If
        End If

    End Sub

    Private Sub MainPicture_Click(ByVal sender As System.Object, ByVal e As MouseEventArgs) Handles MainPicture.MouseDown

        Dim X As Integer = Math.Floor(e.X / TILE_WIDTH)
        Dim Y As Integer = Math.Floor(e.Y / TILE_WIDTH)

        If Not Map.IsOnMap(X, Y) Then Exit Sub

        Dim Cell As Integer = Cells.GetId(X, Y)

        If EditionType = 0 And (Modfrm.BrushNormal.Checked Or Modfrm.BrushContinue.Checked) Then

            If e.Button = Windows.Forms.MouseButtons.Left Then
                PaintOnCell(Cell, False)
            ElseIf e.Button = Windows.Forms.MouseButtons.Right Then
                PaintOnCell(Cell, True)
            End If

        ElseIf EditionType = 1 Then

            If Accessfrm.AccessGlobal.Checked Then
                If e.Button = Windows.Forms.MouseButtons.Left Then
                    Map.Accessibility(Cell) += 1
                    If Map.Accessibility(Cell) > 2 Then Map.Accessibility(Cell) = 0
                Else
                    Map.Accessibility(Cell) -= 1
                    If Map.Accessibility(Cell) < 0 Then Map.Accessibility(Cell) = 2
                End If
            Else
                If e.Button = Windows.Forms.MouseButtons.Left Then
                    Dim Bool As Integer = 0
                    If Accessfrm.DoorTop.Checked Then Bool += 1
                    If Accessfrm.DoorBottom.Checked Then Bool += 2
                    If Accessfrm.DoorLeft.Checked Then Bool += 4
                    If Accessfrm.DoorRight.Checked Then Bool += 8
                    Map.DoorAccess(Cell) = Bool
                Else
                    Map.DoorAccess(Cell) = 0
                End If
            End If
            DrawAll()

        End If

    End Sub

    Private Sub InitMap()

        Map.Init()
        DrawAll()

    End Sub

    Private Sub PaintOnCell(ByVal Cell As Integer, ByVal [Erase] As Boolean)

        Dim Draw As Boolean = False

        Dim i As Integer = 0
        Dim StartPoint As Point = Cells.GetPos(Cell)

        For Each Selected As Point In If([Erase], New Point() {StartPoint}, Modfrm.SelectedPoints)

            Dim ActualPoint As Point = If([Erase], StartPoint, StartPoint + Selected)

            If ActualPoint.X < MAP_WIDTH AndAlso ActualPoint.Y < MAP_HEIGHT Then

                Cell = Cells.GetId(ActualPoint.X, ActualPoint.Y)

                Dim ToWrite As Integer = If([Erase], -1, Modfrm.SelectedCells(i))
                Dim Tiles As Dictionary(Of Integer, Integer) = Nothing
                If Modfrm.ModificationHigh.Checked Then
                    Tiles = Map.MapTilesHigh
                ElseIf Modfrm.ModificationMiddle.Checked Then
                    Tiles = Map.MapTilesMiddle
                ElseIf Modfrm.ModificationLow.Checked Then
                    Tiles = Map.MapTilesLow
                End If

                If Not Tiles Is Nothing AndAlso Tiles.ContainsKey(Cell) AndAlso Not Tiles(Cell) = ToWrite Then
                    Tiles(Cell) = ToWrite
                    Draw = True
                End If

            End If

            i += 1

        Next

        If Draw Then DrawAll()

    End Sub

    Private Sub DrawAll()

        Dim b As New Bitmap(TOTAL_WIDTH + TILE_WIDTH, TOTAL_HEIGHT + TILE_WIDTH)
        Dim g As Graphics = Graphics.FromImage(b)

        If Couche1ToolStripMenuItem.Checked Then Map.DrawLow(g)
        If Couche2ToolStripMenuItem.Checked Then Map.DrawMiddle(g)
        If Couche3ToolStripMenuItem.Checked Then Map.DrawHigh(g)
        If GrilleToolStripMenuItem.Checked Then Map.DrawGrid(g)
        If EditionType = 1 Then Map.DrawAccess(g)

        MainPicture.Image = b

    End Sub

    Private Sub ExitProgram()
        Application.Exit()
    End Sub

    Private Sub EnregistrerSousToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles EnregistrerSousToolStripMenuItem.Click

        Dim Save As New SaveFileDialog
        Save.Filter = "Cartes RPG (*.mrm)|*.mrm"
        If Save.ShowDialog = Windows.Forms.DialogResult.OK Then
            FileName = Save.FileName
            Map.ToFile(FileName)
        End If

    End Sub

    Private Sub OuvrirToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles OuvrirToolStripMenuItem.Click

        Dim Open As New OpenFileDialog
        Open.Filter = "Cartes RPG (*.mrm)|*.mrm"
        If Open.ShowDialog = Windows.Forms.DialogResult.OK Then
            Map.FromFile(Open.FileName)
        End If
        DrawAll()

    End Sub

    Private Sub EnregistrerToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles EnregistrerToolStripMenuItem.Click

        If FileName = "" Then
            EnregistrerSousToolStripMenuItem_Click(sender, e)
        Else
            Map.ToFile(FileName)
        End If

    End Sub

    Private Sub TerrainToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles TerrainToolStripMenuItem.Click
        EditionType = 0
        DrawAll()
    End Sub

    Private Sub AccessibiliteToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles AccessibilitéToolStripMenuItem.Click
        EditionType = 1
        DrawAll()
    End Sub

    Private Sub EditionToolStripMenuItem1_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles EditionToolStripMenuItem1.Click
        Modfrm.Show()
    End Sub

    Private Sub AccessibilitéToolStripMenuItem1_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles AccessibilitéToolStripMenuItem1.Click
        Accessfrm.Show()
    End Sub
End Class

File: Soutenance 3/MapEditorReborn/MapEditorReborn/Mainfrm.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Windows.Forms;

namespace MapEditorReborn
{
    public partial class Mainfrm : Form
    {
        private Modfrm modfrm;
        public Map Map;
        private Accessfrm accessfrm;
        public Tiles Tiles = new Tiles();

        public int EditionType = 0;

        private bool painting;
        private string fileName;
        private int lastCell;

        public Mainfrm()
        {
            InitializeComponent();
            painting = false;
            fileName = "";
            lastCell = -1;
            modfrm = new Modfrm();
            accessfrm = new Accessfrm();
            Map = new Map(this);
        }


        private void MainFrm_Load(object sender, EventArgs e)
        {
            InitMap();
        }

        private void NouveauToolStripMenuItem_Click(object sender, EventArgs e)
        {
            InitMap();
        }

        private void QuitterToolStripMenuItem_Click(object sender, EventArgs e)
        {
            ExitProgram();
        }

        private void GrilleToolStripMenuItem_Click(object sender, EventArgs e)
        {
            DrawAll();
        }

        private void Couche1ToolStripMenuItem_Click(object sender, EventArgs e)
        {
            DrawAll();
        }

        private void Couche2ToolStripMenuItem_Click(object sender, EventArgs e)
        {
            DrawAll();
        }

        private void Couche3ToolStripMenuItem_Click(object sender, EventArgs e)
        {
            DrawAll();
        }

        private void MainPicture_MouseUp(object sender, MouseEventArgs e)
        {
            painting = false;
        }

        private void MainPicture_MouseDown(object sender, MouseEventArgs e)
        {
            painting = true;
        }

        private void MainPicture_MouseMove(object sender, MouseEventArgs e)
        {
            int x = (int)Math.Floor((double)e.X / Constants.TileWidth) - Cells.DeltaScrolling.X;
            int y = (int)Math.Floor((double)e.Y / Constants.TileWidth) - Cells.DeltaScrolling.Y;

            if (!Map.IsOnMap(x, y))
                return;

            int cell = Cells.GetId(x, y);

            if (cell != lastCell)
            {
                lastCell = cell;

                LabelStatus.Text = "X : " + x + "  Y: " + y + "  Id: " + cell + " DeltaScrolling: " + Cells.DeltaScrolling;

                if (EditionType == 0)
                {
                    if (painting && modfrm.BrushContinue.Checked)
                    {
                        if (e.Button == MouseButtons.Left)
                            PaintOnCell(cell, false);
                        else if (e.Button == MouseButtons.Right)
                            PaintOnCell(cell, true);
                    }
                }
            }
        }

        private void MainPicture_MouseClick(object sender, MouseEventArgs e)
        {
            int x = (int)Math.Floor((double)e.X / Constants.TileWidth) - Cells.DeltaScrolling.X;
            int y = (int)Math.Floor((double)e.Y / Constants.TileWidth) - Cells.DeltaScrolling.Y;

            if (!Map.IsOnMap(x, y))
                return;

            int cell = Cells.GetId(x, y);

            if (EditionType == 0 && (modfrm.BrushNormal.Checked || modfrm.BrushContinue.Checked))
            {
                if (e.Button == MouseButtons.Left)
                    PaintOnCell(cell, false);
                else if (e.Button == MouseButtons.Right)
                    PaintOnCell(cell, true);

            }
            else if (EditionType == 1)
            {
                if (accessfrm.AccessGlobal.Checked)
                {
                    if (e.Button == MouseButtons.Left)
                    {
                        Map.Accessibility[cell] += 1;
                        if (Map.Accessibility[cell] > 2)
                            Map.Accessibility[cell] = 2;
                    }
                    else
                    {
                        Map.Accessibility[cell] -= 1;
                        if (Map.Accessibility[cell] < 0)
                            Map.Accessibility[cell] = 2;
                    }
                }
                else
                {
                    if (e.Button == MouseButtons.Left)
                    {
                        int boolean = 0;
                        if (accessfrm.DoorTop.Checked)
                            boolean += 1;
                        if (accessfrm.DoorBottom.Checked)
                            boolean += 2;
                        if (accessfrm.DoorLeft.Checked)
                            boolean += 4;
                        if (accessfrm.DoorRight.Checked)
                            boolean += 8;
                        Map.SideAccess[cell] = boolean;
                    }
                    else
                        Map.SideAccess[cell] = 0;
                }
                DrawAll();
            }
        }

        private void InitMap()
        {
            Map.Init();
            DrawAll();
        }

        private void PaintOnCell(int cell, bool erase)
        {
            bool draw = false;
            int i = 0;
            Point startPoint = Cells.GetPos(cell);

            foreach (Point selected in (erase ? new List<Point> {startPoint} : modfrm.SelectedPoints))
            {
                Point actualPoint = erase ? startPoint :  new Point(startPoint.X + selected.X, startPoint.Y + selected.Y);

                if (actualPoint.X < Constants.MapWidth && actualPoint.Y < Constants.MapHeight)
                {
                    cell = Cells.GetId(actualPoint.X, actualPoint.Y);
                    int toWrite = erase ? -1 : modfrm.SelectedCells[i];
                    Dictionary<int, int> tiles = null;

                    if (modfrm.ModificationHigh.Checked)
                        tiles = Map.MapTilesHigh;
                    else if (modfrm.ModificationMiddle.Checked)
                        tiles = Map.MapTilesMiddle;
                    else if (modfrm.ModificationLow.Checked)
                        tiles = Map.MapTilesLow;

                    if (tiles != null && tiles.ContainsKey(cell) && tiles[cell] != toWrite)
                    {
                        tiles[cell] = toWrite;
                        draw = true;
                    }
                }
                i++;
            }
            if (draw)
                DrawAll();
        }

        public void DrawAll()
        {
            if (MainPicture.Image != null)
                MainPicture.Image.Dispose();

            Bitmap b = new Bitmap(Constants.TotalWidth + Constants.TileWidth, Constants.TotalHeight + Constants.TileWidth);
            Graphics g = Graphics.FromImage(b);

            if (Couche1ToolStripMenuItem.Checked)
                Map.DrawLow(g);
            if (Couche2ToolStripMenuItem.Checked)
                Map.DrawMiddle(g);
            if (Couche3ToolStripMenuItem.Checked)
                Map.DrawHigh(g);
            if (GrilleToolStripMenuItem.Checked)
                Map.DrawGrid(g);
            if (EditionType == 1)
                Map.DrawAccess(g);
            MainPicture.Image = b;

            g.Dispose();
        }

        private void ExitProgram()
        {
            Application.Exit();
        }

        private void EnregistrerSousToolStripMenuItem_Click(object sender, EventArgs e)
        {
            SaveFileDialog save = new SaveFileDialog();
            save.Filter = "Cartes RPG (*.mrm)|*.mrm";
            if (save.ShowDialog() == DialogResult.OK)
            {
                fileName = save.FileName;
                Map.ToFile(fileName);
            }
        }

        private void OuvrirToolStripMenuItem_Click(object sender, EventArgs e)
        {
            OpenFileDialog open = new OpenFileDialog();
            open.Filter = "Cartes RPG (*.mrm)|*.mrm";
            if (open.ShowDialog() == DialogResult.OK)
                Map.FromFile(open.FileName);
            DrawAll();
        }

        private void EnregistrerToolStripMenuItem_Click(object sender, EventArgs e)
        {
            if (fileName ==  "")
                EnregistrerSousToolStripMenuItem_Click(sender, e);
            else
            {
                Map.ToFile(fileName);
            }
        }

        private void TerrainToolStripMenuItem_Click(object sender, EventArgs e)
        {
            EditionType = 0;
            DrawAll();
        }

        private void AccessibilitéToolStripMenuItem_Click(object sender, EventArgs e)
        {
            EditionType = 1;
            DrawAll();
        }

        private void EditionToolStripMenuItem1_Click(object sender, EventArgs e)
        {
            if (modfrm == null || modfrm.IsDisposed)
                modfrm = new Modfrm();
            modfrm.Show();
        }

        private void AccessibilitéToolStripMenuItem1_Click(object sender, EventArgs e)
        {
            if (accessfrm == null || accessfrm.IsDisposed)
                accessfrm = new Accessfrm();
            accessfrm.Show();
        }

        private void MainfrmKeyDown(object sender, KeyEventArgs e)
        {
            Point aux = new Point();

            if (e.KeyCode == Keys.Left)
                aux = new Point(-1, 0);
            else if (e.KeyCode == Keys.Right)
                aux = new Point(1, 0);
            else if (e.KeyCode == Keys.Up)
                aux = new Point(0, -1);
            else if (e.KeyCode == Keys.Down)
                aux = new Point(0, 1);

            if (aux.X == 0 && aux.Y == 0)
                return;

            Cells.DeltaScrolling = 
                new Point(Cells.DeltaScrolling.X + aux.X, 
                    Cells.DeltaScrolling.Y + aux.Y);

            DrawAll();
        }
    }
}
